src/env.cc

```diff
@@ -299,6 +372,145 @@ void Environment::CreateProperties() {
+      return;                                                                 \
+    }                                                                         \
+    PropertyName##_.Reset(isolate_, field);                                   \
+  } while (0);
+  ENVIRONMENT_STRONG_PERSISTENT_TEMPLATES(V)
+#undef V
+
+#define V(PropertyName, TypeName)                                             \
+  do {                                                                        \
+    Local<TypeName> field;                                                    \
+    if (!snapshot_data->ReadObject<TypeName>(context,                         \
+            SnapshotReadData::kAllowEmpty).To(&field)) {                      \
+      return;                                                                 \
+    }                                                                         \
+    PropertyName##_.Reset(isolate_, field);                                   \
+  } while (0);
+  ENVIRONMENT_STRONG_PERSISTENT_VALUES(V)
+#undef V
+
+  uint64_t expected_native_module_count;
+  if (!snapshot_data->ReadUint64().To(&expected_native_module_count)) return;
+  for (uint64_t i = 0; i < expected_native_module_count; i++) {
+    std::string str;
+    if (!snapshot_data->ReadString().To(&str)) return;
+    native_modules_in_snapshot.insert(str);
+  }
+
+  uint64_t expected_base_object_count;
+  if (!snapshot_data->ReadUint64().To(&expected_base_object_count)) return;
+
+#ifdef DEBUG
+  std::set<const BaseObject*> deserialized_objects;
+#endif
+  for (uint64_t i = 0; i < expected_base_object_count; i++) {
+    std::string name;
+    if (!snapshot_data->StartReadEntry(nullptr).To(&name)) return;
+    BaseObject* obj = BaseObjectDeserializer::Deserialize(
+        name, this, snapshot_data);
+    if (obj == nullptr) return;
+#ifdef DEBUG
+    deserialized_objects.insert(obj);
+#endif
+  }
+
+#ifdef DEBUG
+  // Make sure all BaseObject instances have been attached to this Environment.
+  ForEachBaseObject([&](const BaseObject* obj) {
+    deserialized_objects.erase(obj);
+  });
+  CHECK(deserialized_objects.empty());
+#endif
+
+  Local<Context> ctx;
+  if (!snapshot_data->ReadObject<Context>(context).To(&ctx)) return;
+  if (ctx != context)
+    snapshot_data->add_error("Context mismatch");
+  snapshot_data->EndReadEntry();
+}
+
+void Environment::Serialize(SnapshotCreateData* snapshot_data) const {
+  snapshot_data->StartWriteEntry("Environment");
+  async_hooks()->Serialize(snapshot_data);
+  immediate_info()->Serialize(snapshot_data);
+  tick_info()->Serialize(snapshot_data);
+  stream_base_state_.Serialize(snapshot_data);
+  should_abort_on_uncaught_toggle_.Serialize(snapshot_data);
+  performance_state()->Serialize(snapshot_data);
+
+  snapshot_data->WriteBool(has_run_bootstrapping_code());
+  snapshot_data->WriteBool(can_call_into_js());
+
+  snapshot_data->WriteUint32(module_id_counter_);
```

Comment:
What's the rationale behind doing WriteBool/WriteUint32 etc. instead of serializing C++ structures into the code being emitted? Low-level writers makes sense for V8 because they emit assembly for the blob, but we generate C++ files with static data for these blobs and I think we should take advantage of that. These kinds of low-level writers and readers can be quite annoying and make it more difficult for new contributors to understand and debug crashes/mismatch, from my experience.

Reply:
> What's the rationale behind doing WriteBool/WriteUint32 etc. instead of serializing C++ structures into the code being emitted? Low-level writers make sense for V8 because they emit assembly for the blob, but we generate C++ files with static data for these blobs and I think we should take advantage of that.

I’m not entirely sure if that’s what you mean, but if you suggest emitting code in the snapshot file that creates these objects on process startup, then that’s something I’d really like to avoid.I don’t think it would be significantly easier to do that. But more importantly, with this approach things like userland snapshotting can eventually become possible, and that’s definitely a major goal here for me.

> These kinds of low-level writers and readers can be quite annoying and make it more difficult for new contributors to understand and debug crashes/mismatch, from my experience.

I’ve thought about JSON as an alternative, but that ultimately seemed like it would come with overhead that we’d want to avoid. I’ve done my best here to provide a good debugging experience, though, and I while obviously can’t speak for new contributors, it enabled me to rather easily figure out mismatches. And as far as crashes go, I think the only ones that I experienced came from reading empty handles from V8 and then accessing them, but the API here should be good enough to handle that by now imo.

I think writing out C++ structures into the snapshot would provide a much worse debugging experience, too.

Reply:
> if you suggest emitting code in the snapshot file that creates these objects on process startup

Sort of, but not really - I am talking about emitting the data as static, plain objects being passed into initializers, like what we do for generating the map of builtin JS code. Then there will be an intermediate representation for a static graph of data that will be further unfolded into the runtime during startup. Then instead of wondering "whether the env->some_count() is X", you'll just see in the generated code that X is placed in the initializer of a structure, which will be later unfolded into the Environment with firstly `static_data = { X... }` in the generated code, and then`env->set_some_count(static_data.count)`, where the shape of `static_data` is predefined and can be understood by the debugger automatically.

> But more importantly, with this approach things like userland snapshotting can eventually become possible, and that’s definitely a major goal here for me.

I don't see how those are contradicting with each other? We could have, say, a static instance of a well-defined class whose fields are initialized in a human-readable way in the generated C++ code for bootstrapping the built-in instance, as well as dynamic instances of that class that can be mapped from contiguous blobs at run time. Then at least when debugging that builtin blob what's in there is already visible in the C++ file, and can be displayed nicely in a human readable format with a native debugger (since they'll be associated with the right symbols at program startup).

If you can only see whether there's something wrong as you unfold these bytes into the Environment, it could be quite confusing once there are loops/conditionals for optional/variable-sized data - then you'd need to follow through the unfolding to figure out what's wrong, and remember what bytes come after what - because the place where it crashes can be far away from where it starts to go wrong, and having side-effects while you unfold would make that even more difficult to figure out.

Reply:
> If you can only see whether there's something wrong as you unfold these bytes into the Environment, it could be quite confusing once there are loops/conditionals for optional/variable-sized data - then you'd need to follow through the unfolding to figure out what's wrong, and remember what bytes come after what - because the place where it crashes can be far away from where it starts to go wrong, and having side-effects while you unfold from that raw, contiguous blob would make that even more difficult to figure out.

Just in case you missed it, there are typechecking and scoping tags in the generated snapshot data, which have been very effective for me in avoiding situations like that – while working on this, whenever I had an issue like this, it was always caught very early when serialization and deserialization were out of sync in some way.

I like the idea of having a human-readable format of the snapshot data. I’ve added a `--dump-snapshot` here, that seems like it could be helpful. When deserialization doesn’t match the expected types, both a description of the mismatch is given, and the index where the error occurred, which can be pretty easily looked up in the `node --dump-snapshot` output.

For example, if I “accidentally” forget writing the base object count, this is the output when trying to deserialize:

```
$ ./node -p 42
Snapshot error: At [5198] Environment:BaseObjects: Unexpected tag kEntryStart (expected kUint64)
Snapshot error: At [5198] Environment:BaseObjects: Entries left on snapshot stack
 1: 0x564833e5d9df node::DumpBacktrace(_IO_FILE*) [./node]
 2: 0x564833f26efa node::Abort() [./node]
 3: 0x5648340698cc node::SnapshotDataBase::PrintErrorsAndAbortIfAny() [./node]
 4: 0x564833f98bd0 node::NodeMainInstance::CreateMainEnvironment(int*) [./node]
[…]
Aborted (core dumped)
$ ./node --dump-snapshot
[…]
5115     String: "internal/validators"
5144     String: "path"
5158     String: "timers"
5174     EndEntry
5175   StartEntry: [BaseObjects]
5197     StartEntry: [NoBindingData]
5221       Object index: 32
5231       EndEntry
5232     EndEntry
5233   Object index: 33
5243   EndEntry
```
I don’t think you’d get that kind of debugging from C++ compiler errors, tbh.

> Since we are adding support for built-in support first, why not just start with static initializers first, and add dynamic initializers later?

That seems like a lot of extra complexity to me, especially since dynamic storage is something that I’d implement anyway and I still don’t quite see concrete advantages to writing static objects.

Reply:
> there are typechecking and scoping tags in the generated snapshot data

Yes, I've noticed them, but still it requires unfolding the data on-the-fly to make use of them (or I guess you can try to search for tags in a blob of bytes and map them mentally with prior knowledge about the encoding, but that sounds like poor-man's DWARF here and not quite straightforward)

> I don’t think you’d get that kind of debugging from C++ compiler errors, tbh.

Why not? I think the equivalent would just be errors for forgetting to initialize one of the fields in a initializer and other uncompilable things. If there is a predefined structure that describe a set of plain data that need to be unfolded into the Environment, in the debugger you can easily just inspect instances of them and do evaluations on them (and that'll support whatever GUI debugger for that matter). The `--dump-snapshot` sounds like a good idea, but I can't see why they make these direct low-level readers/writers a must. I think it'd be even easier to implement it with a predefined structure. Then you can either use the debugger to print a structure on-the-fly, because that encoding is already built into the binary by the debugger, or just decode each field of that structure at one go to stdout for a command line option.

Reply:
> > I don’t think you’d get that kind of debugging from C++ compiler errors, tbh.
> 
> Why not? I think the equivalent would just be errors for forgetting to initialize one of the fields in a initializer and other uncompilable things.

Right, but that doesn't always give you helpful error messages.

> If there is a predefined structure that describe a set of plain data that need to be unfolded into the Environment, in the debugger you can easily just inspect instances of them and do evaluations on them (and that'll support whatever GUI debugger with nesting for that matter). That's what you can do with `node::native_module::NativeModuleLoader::instance_.source_`, for example.

That requires that you have a debug build of Node.js and are familiar with how to use debuggers -- since you mentioned ease of use for newcomers, that seems to be a counterargument? To "easily just" do something with a native debugger is a big jump for some people. I also don't see how it's easier than printing the data in a structured format like `--dump-snapshot` does.

> The `--dump-snapshot` sounds like a good idea, but I can't see why they make these direct low-level readers/writers a must.

As I mentioned above, supporting userland snapshotting is a goal here, and that kind of requires a sensible snapshotting format that does not require extra compilation steps like using C++ would.

Generally, I'm not really willing to re-write the code in this PR to something that I believe makes less sense with the knowledge that I'd have to re-write it back to the current state afterwards. That would be a huge waste of time on my end. 

> I think it'd be even easier to implement it with a predefined structure. Then you can either use the debugger to print a structure on-the-fly, because that encoding is already built into the binary by the debugger, or just decode each field of that structure at one go to stdout for a command line option.

To be honest, that doesn't sound easier at all to me, especially manually decoding each field of a structure like that.

Reply:
The userland snapshotting is definitely a key goal that we'd like to work towards. The idea is to improve startup time through the *full* bootstrap cycle including entry point userland code. For use case, consider a serverless environment where the Node.js instance loads a fairly significant amount of userland library code at startup. Moving to an approach that requires recompilation of the Node.js binary is a non-starter.

Reply:
> Right, but that doesn't always give you helpful error messages.

In most cases, I think that'll be helpful enough - a message like `Unexpected tag kEntryStart (expected kUint64)` would not be easy to understand for someone who's not familiar with the snapshot layout either. In addition, error messages emitted on readable texts should be somewhat easier to digest and reason about than error messages emitted on a chuck of raw bytes that can only be turned into a human-readable format with a working Node.js build (what happens if `--dump-snapshot` also crash?).

> That requires that you have a debug build of Node.js and are familiar with how to use debuggers -- since you mentioned ease of use for newcomers, that seems to be a counterargument?

I don't think that'll require a debug build? The release build is compiled with -Os , and the data being static means there should be enough symbols for one to decode these structures (at least I've been able to inspect `node::native_module::NativeModuleLoader::instance_.source_` with the release build so far). When I mention the ease of use for newcomers, I was talking about a scenario when someone who's not familiar with Node.js (but familiar with native debuggers) trying to debug the crash. A `--dump-snapshot` option can be useful alongside debuggable symbols (although still it requires mental effort to understand the snapshot layout, and text processing techniques given the size of the output), but I don't see them at odds? Native debuggers would help you understand what actually happens at runtime as the code unfolds the data, while `--dump-snapshot` tells you how Node.js would interpret the data (but it's not certain, because it's detached from the actual deserialization code making side-effects). These two can co-exist, and for people who can use native debuggers (I suppose there would be more of them compared to people who understand Node.js internals, or it's much easier to find tutorials online for native debugging) they'll just have an additional tool.

> that kind of requires a sensible snapshotting format that does not require extra compilation steps like using C++ would.

Why would that be a requirement? Perhaps I did not make myself clear when I explained about it. The idea is, for serialization, it goes through

1. Picking up useful data from an Environment into a plain object with a predefined structure. To add additional data one would likely have to change this structure and compile error would be thrown if the builtin snapshot doesn't match it (but not for dynamic snapshots).
2. Serialize the object capturing Environment states into either a C++ file that can be compiled into the binary - which will be done as what's written in snapshot_builder.cc, or a chunk of bytes that can be deserialized - which will be done by the writers as implemented in this PR.

For deserialization, it goes through

1. Deserialize the data into a plain object with a predefined structure - for the builtin snapshot, this would be done automatically by the program loader, no memory access or copying is necessary, checking would be minimal since there are already checks at compile time. For dynamic snapshots, this would be done by the readers as implemented in this PR and more checks, memory access and copying would then be inevitable. Note that this step would be side-effect free, making it easier to debug crashes.
2. Unfold the plain object into the Environment - this is where the side effect happens. If Node.js is run with `--dump-snapshot`, then instead of unfolding we'd just print the fields of the structure one by one (but the printer code would be fairly straight forward as you don't need to worry about who's a boolean or who's an uint32, etc. at this point).

So there's no conflict for dynamic snapshot here, they'll simply be based on a different serializer and deserializer under the hood. The intermediate plain object can be inspected in the debugger and `--dump-snapshot` can also be implemented with a different deserializer.

> Generally, I'm not really willing to re-write the code in this PR to something that I believe makes less sense with the knowledge that I'd have to re-write it back to the current state afterwards. That would be a huge waste of time on my end.

As explained above I don't think the current low-level reader/writer code would need to be reverted for support of dynamic snapshot, but this brings up an interesting question - I've been working towards that direction in my prototype, and have been careful to preserve certain invariants being violated here. Now it may not be intentional but this PR sort of turns things into a competition in which the design gets materialized first will win and the code written for the other direction would be another waste of time (even though in the case of serialization/deserialization, the designs aren't at odd with each other, but still code would be wasted), so essentially it would push towards design decisions based on speed of implementation on one end, instead of collaboration on multiple ends, which would be unfortunate.

> To be honest, that doesn't sound easier at all to me, especially manually decoding each field of a structure like that.

Why not? As an example, to print the counters, you'd have something like

```
// In the deserializer to the plain object for dynamic snapshot
// (not necessary for builtin snapshot - the program loader will do this automatically)
raw_snapshot_data->ReadBool().To(&snapshot_data.module_id_counter);
raw_snapshot_data-> ReadUint32().To(&snapshot_data.module_id_counter);
raw_snapshot_data-> ReadUint32().To(&snapshot_data.module_id_counter);

// In the printer - we don't necessarily need to manually repeat these, it
// can be done with a macro that is used to both declare these fields and to print them out.
out << "module_id_counter: " << snapshot_data.module_id_counter << "\n";
out << "script_id_counter: " << snapshot_data.script_id_counter << "\n";
out << "function_id_counter: " << snapshot_data. unction_id_counter << "\n";
```

Reply:
> Moving to an approach that requires recompilation of the Node.js binary is a non-starter.

@jasnell I agree with that, but as explain in the previous comment, that's not what my suggestion implies - this will just require two different serializer/deserializer for the builtin snapshot and the dynamic snapshot, but adds additional debugging capability to and reduces memory access and copying for the builtin snapshot.

Reply:
> > Right, but that doesn't always give you helpful error messages.
> 
> In most cases, I think that'll be helpful enough - a message like `Unexpected tag kEntryStart (expected kUint64)` would not be easy to understand for someone who's not familiar with the snapshot layout either.

If you feel that the error messages could be improved here, I’m happy to work on that. But generally, these messages will only pop up when somebody is modifying the snapshot code, which should already be a pretty good start for us.

> In addition, error messages emitted on readable texts should be somewhat easier to digest and reason about than error messages emitted on a chuck of raw bytes that can only be turned into a human-readable format with a working Node.js build

I don’t think that’s necessarily true, so long as the error messages are clear about what the problem is.

> (what happens if `--dump-snapshot` also crash?).

I mean, I guess that could happen, but it doesn’t seem like a problem that would occur frequently enough in real-world usage to worry about? The code here is written to be relatively robust against crashes, I’d say, with relatively simple structures used only, enforced explicit failure checking and memory access to the snapshot blob only performed through a single pair of read/write functions with bounds checking.

> > That requires that you have a debug build of Node.js and are familiar with how to use debuggers -- since you mentioned ease of use for newcomers, that seems to be a counterargument?
> 
> I don't think that'll require a debug build? The release build is compiled with -Os , and the data being static means there should be enough symbols for one to decode these structures (at least I've been able to inspect `node::native_module::NativeModuleLoader::instance_.source_` with the release build so far). 

Yeah, it looks like that works :+1:

> > that kind of requires a sensible snapshotting format that does not require extra compilation steps like using C++ would.
> 
> Why would that be a requirement? Perhaps I did not make myself clear when I explained about it. The idea is, for serialization, it goes through
> 
>     1. Picking up useful data from an Environment into a plain object with a predefined structure. To add additional data one would likely have to change this structure and compile error would be thrown if the builtin snapshot doesn't match it (but not for dynamic snapshots).
> 
>     2. Serialize the object capturing Environment states into either a C++ file that can be compiled into the binary - which will be done as what's written in snapshot_builder.cc, or a chunk of bytes that can be deserialized - which will be done by the writers as implemented in this PR.
> 
> 
> For deserialization, it goes through
> 
>     1. Deserialize the data into a plain object with a predefined structure - for the builtin snapshot, this would be done automatically by the program loader, no memory access or copying is necessary, checking would be minimal since there are already checks at compile time. For dynamic snapshots, this would be done by the readers as implemented in this PR and more checks, memory access and copying would then be inevitable. Note that this step would be side-effect free, making it easier to debug crashes.
> 
>     2. Unfold the plain object into the Environment - this is where the side effect happens. If Node.js is run with `--dump-snapshot`, then instead of unfolding we'd just print the fields of the structure one by one (but the printer code would be fairly straight forward as you don't need to worry about who's a boolean or who's an uint32, etc. at this point).
> 
> 
> So there's no conflict for dynamic snapshot here, they'll simply be based on a different serializer and deserializer under the hood. The intermediate plain object can be inspected in the debugger and `--dump-snapshot` can also be implemented with a different deserializer.
> 

Okay, but that seems like at least twice the work, and more than twice the code to maintain? First serialize into plain objects, then serialize again into C++ code or a byte array? You say “simply based on a different serializer and deserializer”, but that would still be two different serializer and deserializer implementations, plus the overhead of defining the plain object structures?

> > Generally, I'm not really willing to re-write the code in this PR to something that I believe makes less sense with the knowledge that I'd have to re-write it back to the current state afterwards. That would be a huge waste of time on my end.
> 
> As explained above I don't think the current low-level reader/writer code would need to be reverted for support of dynamic snapshot, but this brings up an interesting question - I've been working towards that direction in my prototype, and have been careful to preserve certain invariants being violated here.

When you say that certain invariants are being violated here, it would be helpful if you could spell out what those are. Otherwise there’s not much I can do about it.

> Now it may not be intentional but this PR sort of turns things into a competition in which the design gets materialized first will win and the code written for the other direction would be another waste of time (even though in the case of serialization/deserialization, the designs aren't at odd with each other, but still code would be wasted), so essentially it would push towards design decisions based on speed of implementation on one end (which can be affected by many factors other than the design itself), instead of collaboration on multiple ends, which would be unfortunate.

I’m not sure if this is what you’re saying, but I don’t think it’s there’s a first-implementation-wins scenario here, given that we do update our code regularly if requirements change or we see improvements to be made. If there were, for example, an existing implementation for snapshotting the `Environment` using plain C++ objects, I would have opened a PR that either augments or replaces that implementation, depending on what that previous implementation would have looked like.

Generally, no, it’s not intentional of me to push a specific design here, and certainly not by opening a PR with one design before another one is being opened, and nor do I believe that the code that is being suggested here is not going to be replaced or augmented in some way. I do genuinely believe that the current state of this PR does strike a better balance between peformance, code complexity, ease of debugging, and future extensibility than a hypothetical plain C++ object implementation would. 

> > To be honest, that doesn't sound easier at all to me, especially manually decoding each field of a structure like that.
> 
> Why not? As an example, to print the counters, you'd have something like
> 
> ```
> // In the deserializer to the plain object for dynamic snapshot
> // (not necessary for builtin snapshot - the program loader will do this automatically)
> raw_snapshot_data->ReadBool().To(&snapshot_data.module_id_counter);
> raw_snapshot_data-> ReadUint32().To(&snapshot_data.module_id_counter);
> raw_snapshot_data-> ReadUint32().To(&snapshot_data.module_id_counter);
> 
> // In the printer - we don't necessarily need to manually repeat these, it
> // can be done with a macro that is used to both declare these fields and to print them out.
> out << "module_id_counter: " << snapshot_data.module_id_counter << "\n";
> out << "script_id_counter: " << snapshot_data.script_id_counter << "\n";
> out << "function_id_counter: " << snapshot_data. unction_id_counter << "\n";
> ```

Right, but that’s not really what I meant here – as I mentioned above, you’d still have to write three serialization and deserialization pairs of code here (live objects ↔ plain C++ objects, plain C++ objects ↔ C++ code, plain C++ objects ↔ byte arrays). That’s what I’m talking about when I say it doesn’t seem easier to me.

Reply:
@joyeecheung And to be explicit: I'm not fundamentally opposed to plain C++ objects as an intermediate representation. I'm currently having a hard time seeing how that wouldn't come with significantly increased complexity, but I definitely wouldn't mind being proven wrong about that.

Reply:
> If there were, for example, an existing implementation for snapshotting the Environment using plain C++ objects, I would have opened a PR that either augments or replaces that implementation, depending on what that previous implementation would have looked like.

Hm, it seems unfortunate that [this comment](https://github.com/nodejs/node/pull/30467#discussion_r393249158) went unnoticed, which referenced this prototype based on plain C++ objects in https://github.com/nodejs/node/compare/master...joyeecheung:snapshot-env which was up-to-date with master until mid-March (hence I raised https://github.com/nodejs/node/pull/32538#discussion_r406848282 after the bindings became `BaseObject`s). It includes a snapshot until `bootstrap/loaders.js` is executed (but not `bootstrap/node.js` yet). The `struct EnvSerializeInfo` there is the kind of plain C++ object I talked about. This was what my updates in https://github.com/nodejs/node/issues/17058 were based on (I think I also mentioned [this](https://github.com/joyeecheung/node/commits/snapshot-2) to you at one point, which is a variant of that prototype, but it also exhibits the idea about that `EnvSerializeInfo`). Granted, it does not contain all the Environment fields included in this PR (as it only keeps what's initialized up until loaders are initialized) but It doesn't seem all that complicated to me to add a function that serializes EnvSerializeInfo into a raw chunk of bytes?

BTW, I also noticed that in this PR the AliasedBuffers aren't serialized with the snapshot creator API, which account for quite a few fields being customized into the snapshot data chunk here. The prototype contains [a patch](https://github.com/nodejs/node/commit/7a26d3e58d431c8aebd94d41f590a8250bd8a617) to track the AliasedBuffers with a list (somewhat similar to how BaseObject's are tracked) which allows you to serialize them with the snapshot creator instead of using custom reader/writers. This works without https://chromium-review.googlesource.com/c/v8/v8/+/2144123 because it does not externalize the ArrayBuffers. 

> When you say that certain invariants are being violated here, it would be helpful if you could spell out what those are. 

Most noticeably, having a plain C++ object with structural data that can be inspected using a debugger (like what's been done in `node_javascript.cc` and `node_code_cache.cc`), as well as keeping the template's callback data context-independent (i.e. my review comments).

Reply:
> > If there were, for example, an existing implementation for snapshotting the Environment using plain C++ objects, I would have opened a PR that either augments or replaces that implementation, depending on what that previous implementation would have looked like.
> 
> Hm, it seems unfortunate that [this comment](https://github.com/nodejs/node/pull/30467#discussion_r393249158) went unnoticed,

Fwiw, the current state of this PR does not quite have the problem described in that comment in the sense that `CreateEnvironment()` is broken now.

> which referenced this prototype based on plain C++ objects in [master...joyeecheung:snapshot-env](https://github.com/nodejs/node/compare/master...joyeecheung:snapshot-env) which was up-to-date with master until mid-March (hence I raised [#32538 (comment)](https://github.com/nodejs/node/pull/32538#discussion_r406848282) after the bindings became `BaseObject`s). It includes a snapshot until `bootstrap/loaders.js` is executed (but not `bootstrap/node.js` yet). The `struct EnvSerializeInfo` there is the kind of plain C++ object I talked about.

Okay, yes, that’s helpful for comparison. (Your comment didn’t go unnoticed, but I didn’t know that it had a partial snapshotting implementation of `Environment` included at that point.)

I think the serialization steps in that diff are a good example of what I mean when I say there’s extra complexity with that approach. Yes, it’s relatively straightforward code, but you still have to manually serialize each part of the generated plain C++ object. This PR also comes with some extra code, mostly the stuff in `src/snapshot_support*`, but it bundles the complexity in one place as far as possible. I think that’s an advantage.

> This was what my updates in #17058 were based on (I think I also mentioned [this](https://github.com/joyeecheung/node/commits/snapshot-2) to you at one point, which is a variant of that prototype, but it also exhibits the idea about that `EnvSerializeInfo`). Granted, it does not contain all the Environment fields included in this PR (as it only keeps what's initialized up until loaders are initialized) but It doesn't seem all that complicated to me to add a function that serializes/deserialize EnvSerializeInfo into/from a raw chunk of bytes dynamically?

Probably not, no. I still don’t really see the advantage, but I might try that out tomorrow and see what the diff looks like.

> BTW, I also noticed that in this PR the AliasedBuffers aren't serialized with the snapshot creator API which account for quite a few fields being customized into the snapshot data chunk here.

Err, they *are* serialized with the snapshot creator API? 

> The prototype contains [a patch](https://github.com/nodejs/node/commit/7a26d3e58d431c8aebd94d41f590a8250bd8a617) to track the AliasedBuffers with a list (somewhat similar to how BaseObject's are tracked) which allows you to serialize them with the snapshot creator instead of using custom reader/writers.

I don’t quite get that patch, tbh. I’m sorry if I sound repetitive, but that patch seems to add quite a bit of extra complexity, with the purpose being able to print out all of them? That doesn’t seem to be related to snapshotting?

And as I said, this PR does use the snapshot creator to serialize `ArrayBuffer`, so I’m not really sure what you’re referring to.

> This works without https://chromium-review.googlesource.com/c/v8/v8/+/2144123 because it does not externalize the ArrayBuffers.

Externalizing ArrayBuffers is a no-op at this point, so I’m guessing you’re referring to something else?

> > When you say that certain invariants are being violated here, it would be helpful if you could spell out what those are.
> 
> Most noticeably, having a plain C++ object with structural data that can be inspected using a debugger (like what's been done in `node_javascript.cc` and `node_code_cache.cc`),

Okay, as mentioned above, I’m okay with breaking that invariant.

> as well as keeping the template's callback data context-independent (i.e. my review comments).

Right – I thought you were referring specifically to changes in the scope of this PR.

Reply:
> > This was what my updates in #17058 were based on (I think I also mentioned [this](https://github.com/joyeecheung/node/commits/snapshot-2) to you at one point, which is a variant of that prototype, but it also exhibits the idea about that `EnvSerializeInfo`). Granted, it does not contain all the Environment fields included in this PR (as it only keeps what's initialized up until loaders are initialized) but It doesn't seem all that complicated to me to add a function that serializes/deserialize EnvSerializeInfo into/from a raw chunk of bytes dynamically?
> 
> Probably not, no. I still don’t really see the advantage, but I might try that out tomorrow and see what the diff looks like.

Yeah, I can’t really seem to come up with an implementation idea that doesn’t add a ton of extra code…

I think I’d rather focus on improving the debugging experience here, since that seems to be your main concern (even if I still think that it’s on about the same level as plain C++ object serialization would provide).

Reply:
I’ve pushed an update that should make debugging a bit nicer imo. The error message in case of a failure now contains the context of the mismatch in the snapshot data, e.g. the equivalent of the one mentioned above (forgetting to write the number of `BaseObject`s):

```
Encountered 1 snapshot errors:
At [5448] Environment:BaseObjects: Tried to read object of type kUint64, found type kEntryStart instead
Around:
    5394     String: 'path'
    5408     String: 'timers'
    5424     EndEntry
    5425   StartEntry: [BaseObjects]
*   5447     StartEntry: [NoBindingData]
    5471       Object index: 32
    5481       EndEntry
    5482     EndEntry
    5483   Object index: 33
    5493   EndEntry
(`node --dump-snapshot` dumps the full snapshot data contained in a node binary)
```

It might help if you could make suggestions for what else you would expect from the debugging experience here – both things you think would be provided by C++ error messages and not by this approach, and general ideas.

One obvious such item might be including the names of object fields in the snapshot data. I’m a bit reluctant, given that that would essentially be “dead weight” as far as the actual snapshotting operations are concerned and it’s relatively easy (from my perspect to match the output to the code, but if you feel that that is helpful and should be included, I will do so.

Reply:
> e.g. the equivalent of the one mentioned above (forgetting to write the number of BaseObjects):

For this example alone I think it cannot exist in the C++ object serialization for builtin-snapshots - you'd just serialize it into an array of initializers and the count would not have mattered. The issue with the count would only exist when the number of base objects in the snapshot is not deterministic i.e. as in dynamic snapshots, and here we are giving up what the knowledge of a static blob offers simply because we want to treat it as a dynamic one.

For other types of issues, when there is a type mismatch or when the order in which you read/write some fields is wrong, the compiler would just complain in the way it complains about wrong C++ object initializations. To inspect what's wrong with the output and tinker with it you'd just need a text editor instead of `node --dump-snapshot` or tools like hexdump. The deserialization would be performed by the loader and would not involve so many unaligned memory access/checking/copying caused by the tagging done in the Read* methods.

Reply:
> I think the serialization steps in that diff are a good example of what I mean when I say there’s extra complexity with that approach. Yes, it’s relatively straightforward code, but you still have to manually serialize each part of the generated plain C++ object. This PR also comes with some extra code, mostly the stuff in src/snapshot_support*, but it bundles the complexity in one place as far as possible. I think that’s an advantage.

The actual serialization into C++ is roughly just 80+ lines of code, though, and I don't think it's going to grow that much considering the additional states we need to keep track of outside of the Deserialization/Serialization callbacks of V8 is quite limited (and I don't think it'll involve as much code as required for dynamic snapshot serialization in the snapshot_support* files here since things are fairly deterministic for the builtin snapshot). I think it's worth it for the benefit of inspect-ability of the snapshot blobs at run time. `--dump-snapshot` and the additional dumping at failure are helpful but still limited, and the tagging would add additional burden to people who don't understand the layout of these blobs (compared to layout of a C++ class which would be much more straight forward).

Reply:
>  that patch seems to add quite a bit of extra complexity, with the purpose being able to print out all of them? That doesn’t seem to be related to snapshotting?

It's paired with https://github.com/nodejs/node/commit/c63cc73d2d7abc67cc508359f395a4931fea9286 which will save the snapshot index  of all known AliasedBuffers reachable from the Environment, so that the reference from the Environment can established again at deserialization. Snapshot indexes to AliasedBuffers not directly reachable from the Environment can be saved using the SerializeInternalFieldCallback and the references can be restored during deserialization as well. In addition a check can be added for to make sure we serialize all the AliasedBuffers ever created, or printers help you understand whether the contents of these buffers at serialization/deserialization (because not all of them make sense to be serialized/deserialized, like the arrays that are only used to pass data between C++ and JS. For some, only certain fields of them make sense to be serialized while others need to be erased, like the performance milestones).

Reply:
> > e.g. the equivalent of the one mentioned above (forgetting to write the number of BaseObjects):
> 
> For this example alone I think it cannot exist in the C++ object serialization for builtin-snapshots - you'd just serialize it into an array of initializers and the count would not have mattered. The issue with the count would only exist when the number of base objects in the snapshot is not deterministic i.e. as in dynamic snapshots, and here we are giving up what the knowledge of a static blob offers simply because we want to treat it as a dynamic one.

Right, it’s a contrived example, obviously. The fact that the static snapshot is treated like a dynamic one is kind of the point, though.

> For other types of issues, when there is a type mismatch or when the order in which you read/write some fields is wrong, the compiler would just complain in the way it complains about wrong C++ object initializations. To inspect what's wrong with the output and tinker with it you'd just need a text editor instead of `node --dump-snapshot` or tools like hexdump.

Right, but I don’t see how that is *better* than what the snapshot currently provides in terms of error output. Also, I haven’t had to use something like hex dumping of the data during working on this, the format is quite robust.

> The deserialization would be performed by the loader and would not involve so many unaligned memory access/checking/copying caused by the tagging done in the Read* methods.

I don’t think this is a concern here, or even a fair argument, tbh. The Node.js snapshot blob in this PR is less than 6 KB, compared to 1.6 MB for the V8 snapshot data, and time spent deserializing the V8 snapshot data dwarfs the processing time when reading from the snapshot.

> > I think the serialization steps in that diff are a good example of what I mean when I say there’s extra complexity with that approach. Yes, it’s relatively straightforward code, but you still have to manually serialize each part of the generated plain C++ object. This PR also comes with some extra code, mostly the stuff in src/snapshot_support*, but it bundles the complexity in one place as far as possible. I think that’s an advantage.
> 
> The actual serialization into C++ is roughly just 80+ lines of code, though, and I don't think it's going to grow that much considering the additional states we need to keep track of outside of the Deserialization/Serialization callbacks of V8 is quite limited (and I don't think it'll involve as much code as required for dynamic snapshot serialization in the snapshot_support* files here since things are fairly deterministic for the builtin snapshot).

I’m not disagreeing that *currently* this would be simpler, but it does really not extend to dynamic snapshotting in a comparable way, at least if we also want a good debugging experience for that case. And if you’re saying that extensibility towards dynamic snapshotting is something that this PR should disregard, than I’m strongly disagreeing with that point.

>  I think it's worth it for the benefit of inspect-ability of the snapshot blobs at run time. `--dump-snapshot` and the additional dumping at failure are helpful but still limited, 

That’s why I was asking for examples – how are they more limited?

> and the tagging would add additional burden to people who don't understand the layout of these blobs (compared to layout of a C++ class which would be much more straight forward).

When would that be a typical concern?

Reply:
> > that patch seems to add quite a bit of extra complexity, with the purpose being able to print out all of them? That doesn’t seem to be related to snapshotting?
> 
> It's paired with [c63cc73](https://github.com/nodejs/node/commit/c63cc73d2d7abc67cc508359f395a4931fea9286) which will save the snapshot index of all known AliasedBuffers reachable from the Environment, so that the reference from the Environment can established again at deserialization. 

That’s what this PR does, too, without maintaining an explicit list of them.

> Snapshot indexes to AliasedBuffers not directly reachable from the Environment can be saved using the SerializeInternalFieldCallback and the references can be restored during deserialization as well. 

How is that not way more complex?

> In addition a check can be added for to make sure we serialize all the AliasedBuffers ever created, 

That’s already implicitly checked by V8 through by making sure that all global references are serialized, though.

> or printers help you understand whether the contents of these buffers at serialization/deserialization (because not all of them make sense to be serialized/deserialized, like the arrays that are only used to pass data between C++ and JS. For some, only certain fields of them make sense to be serialized while others need to be erased, like the performance milestones).

Right, that’s a fair point. I’m not really seeing how that’s related specifically to `AliasedBuffer`s, though – it’s a concern that’s valid for all objects, right?

Reply:
> That’s already implicitly checked by V8 through by making sure that all global references are serialized, though.

Yes, but we can print them out in a way we understand.

>  it’s a concern that’s valid for all objects, right?

Yes, but if we can do something to make it easier to inspect AliasedBuffers during serialization/deserialzation, why not? I'd say we could also add facilities for printing base objects as well, to better understand the state of things at snapshot time for debugging, which is also done in that patch, here it can be somewhat replaced by the snapshot dumpers but they currently don't understand the type names of these objects.

Reply:
> here it can be somewhat replaced by the snapshot dumper but it currently doesn't understand the type names of these objects.

Can you explain what the “type name of these objects” is?

Reply:
> type name of these objects

The class names of them. For example, if a `ModuleWrap` appears in the snapshot, how are we supposed to know there is a `ModuleWrap`? (even though for the builtin snapshot, they are not supposed to appear, so we would not add any methods to denote them at the moment)

Reply:
@joyeecheung Okay, yeah, the class names are included in the snapshot currently, both for debugging and for knowing which kind of object we’re supposed to serialize.

Reply:
I think the issue is that these class names are not shared with the current format, and hence for every object in the snapshot, they will be repeated even though this is just structural information for debugging and IMO this information should not be proportional to the number of instances in the snapshot. Setting whether builtin snapshot should be emitted as C++ code aside, having pre-defined structures would reduce the overhead for each C++ object significantly (it's probably 1 uint8_t for the entire object, v.s. number of fields * 8 + number of bytes of the class string + 1 uint8_t for start + 1 uint8_t for end, the latter could be more if we want to nest C++ types and make that nesting debuggable).

Reply:
Generally, I don’t think size is currently an issue for this blob, considering that the corresponding V8 snapshot is ~250 times larger currently. 

If we do think that size becomes an issue eventually, we could probably push the size down quite a bit by optionally encoding 64-bit and 32-bit integers using smaller sizes if they are available, and if necessary use backreferences for things like class names.

> it's probably 1 byte for the entire object

That sounds like an approach where the type information would be encoded as an enum, which I’d definitely want to avoid if at all possible.

Reply:
> That sounds like an approach where the type information would be encoded as an enum, which I’d definitely want to avoid if at all possible.

Why would that be undeseriable? (Referencing my last comments in https://github.com/nodejs/node/pull/32761#discussion_r410085465)

Reply:
@joyeecheung To be clear, I think it’s generally a bad idea to keep global list of all possible cases of something, whether that’s `BaseObject`s or something else. It’s okay if nothing else works, but given that we don’t have to do this, I would not do it at all.